### ISSUE TITLE
[MacOS] Thread.Sleep and await Task.Delay take longer than expected when the app is tabbed out for ~50s

### ISSUE THREAD

**Issue by Liam2349** [xamarin/Xamarin.Forms#4876]
### Description
I have an endless loop running in my app, which sleeps for 500ms between runs, and runs on a manually created background thread. On each iteration, I log the loop time to the console. If the app is in focus, everything is fine - each iteration takes 500ms. If the app is tabbed out for ~50s, the _await Task.Delay(500)_ gets about 10s added to it, taking 10,500ms. It's not 100% consistent - sometimes it's 20s, sometimes it's more like 200ms, but 10s is the most common wait added.

If I tab back into the app, the delay goes back to taking 500ms, as expected, and seems to remain that way until tabbed out again for a period of ~50s.
### Steps to Reproduce

Repro solution here: https://1drv.ms/u/s!Avxe5n2Gio0VyZEYjF4JrqHBIujuuA

GitHub said it's too big to attach, so I'll host it there temporarily, but in the interest of preservation it would be good if one of the Xamarin team members could host it somewhere else. It's just a blank XF project with a Mac target and a tiny bit of code. The repro uses _await Task.Delay_ but the same thing happens with _Thread.Sleep_.

1. Open the solution, run Mac app in Debug
2. Watch the logging messages printed to the output. They show the current loop iteration and the previous loop time.
3. See that the loop time is 500ms.
4. Tab out (click into another window), wait around 50s.
5. Look at the output, and see that the loop time is no longer consistently 500ms. You should see a lot of 10,500ms.

You can see the code in MainPage.xaml.cs inside the shared project. It's just about 30 lines to set up a new thread that runs an endless loop. The loop writes logging messages to the output and to a file on the desktop, for running without the debugger. I used this file to find out that the problem still occurs without the debugger attached.

The code measures the time taken by the delay, using a stopwatch. This time is logged for you to see. When you tab out and wait ~50s, the time is no longer consistently 500ms. I also tested a 50ms wait, and again after tabbing out and waiting ~50s, I was seeing a lot of 10,050ms delays, so there seems to be something special about the 10s being added to it.

I don't know if this is a Mono issue, a XF issue, or a Xamarin.Mac issue. I'm leaning towards it being XF or Xamarin.Mac because this problem doesn't occur until the window is tabbed out.

### Expected Behavior
await Task.Delay and Thread.Sleep should wait for as long as they are told to.

### Actual Behavior
These wait methods will wait longer than told to when the app has been tabbed out for ~50s.

### Basic Information

- Version with issue: XF 3.4.0.1008975
- Last known good version: Unknown
- IDE: VS for Mac 7.7.2 build 21 (Community)
- Platform Target Frameworks:
  - macOS 10.14 (I don't know if it affects other OS versions)
- Nuget Packages: Just Xamarin.Forms
- Affected Devices: macOS 10.14. I am testing this inside a virtual machine but I don't think that's the issue.

### Reproduction Link

Repro solution here: https://1drv.ms/u/s!Avxe5n2Gio0VyZEYjF4JrqHBIujuuA

**Comment by ahmedalejo** [comment]
Pointing to a zip might make it difficult to evaluate even though what you
described is quite easy to reproduce. You could always point to a git repo.

That said, this is most likely to happen in any UI application Android even
like WPF. When most ui apps are backgrounded i.e tabbed, they tend to be
given lesser cpu time which means wait time ou execution speed is reduced
or paused in case of Android Oreo and above, this mostly for battery saving
on mobile and performance reasons on desktop.

You should look into how prioritized background execution is done on Mac,
It might be similar to iOS beginBackgroungFetch; for example on Android,
you’d use a foreground service for this; like when you are playing audio or
something similar that should not be interrupted when backgrounded by
tabbing.

**Comment by Liam2349** [comment]
@ahmedalejo Thanks for the response, but I have never in my life experienced this with a WPF app, nor with an Android app. Android has several ways of running in the background, and for my app I use a foreground service, which I have never seen behave like this.

I google searched for background execution in cocoa as I initially assumed it was some sort of throttling done by the OS, however I could not find one single document on background execution for macOS apps, aside from headless apps, so I don't believe this is the case.

It is also not a case of everything being slowed; as far as I can tell, this is only occurring for _await Task.Delay_ and _Thread.Sleep_. The app that I found this problem with has SignalR running elsewhere and that seems to still be working instantly.

One of my SignalR events will be received by the mac app, update a layout, insert into an SQLite db, put something on the clipboard, and somewhere in this process it will break the 10 second waits caused by the Delay and Sleep methods. Whilst a long wait is in progress, and my SignalR handler does its work, the wait will immediately return and stay as normal for ~50 seconds, as if you had tabbed into the app.

I did check before if I could keep updating some UI to prevent the 10 second delays, but when I kept changing a label's text in every iteration, it didn't help, and the same thing happened. Maybe it's due to the incoming network request, in some manner, though the SignalR heartbeats are not waking it. 

I haven't looked into this further yet, but if there is some kind of throttling done by the OS, I cannot find it documented anywhere, so I presume it to be a Xamarin issue at some level.
